cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/coordinatedwrite/CassandraClusterInfoGroup.java

```diff
@@ -0,0 +1,300 @@
+                                            " and highest=" + last.getRawVersionString());
+        }
+
+        return first.getRawVersionString();
+    }
+
+    @Override
+    public Map<RingInstance, WriteAvailability> clusterWriteAvailability()
+    {
+        if (clusterInfos.size() == 1)
+        {
+            return clusterInfos.get(0).clusterWriteAvailability();
+        }
+
+        Map<String, Map<RingInstance, WriteAvailability>> aggregated = applyOnEach(ClusterInfo::clusterWriteAvailability);
+        Map<RingInstance, WriteAvailability> consolidated = new HashMap<>();
+        aggregated.values().forEach(consolidated::putAll);
+        return consolidated;
+    }
+
+    @Override
+    public Partitioner getPartitioner()
+    {
+        Map<String, Partitioner> aggregated = applyOnEach(ClusterInfo::getPartitioner);
+        Set<Partitioner> partitioners = EnumSet.copyOf(aggregated.values());
+        if (partitioners.size() != 1)
+        {
+            throw new IllegalStateException("Clusters are not running with the same partitioner kind. Found partitioners: " + aggregated);
+        }
+
+        return MapUtils.firstEntry(aggregated).getValue();
+    }
+
+    @Override
+    public void checkBulkWriterIsEnabledOrThrow()
+    {
+        runOnEach(ClusterInfo::checkBulkWriterIsEnabledOrThrow);
+    }
+
+    /**
+     * @return the largest time skew retrieved from the target replicas
+     */
+    @Override
+    public TimeSkewResponse getTimeSkew(List<RingInstance> instances)
+    {
+        if (clusterInfos.size() == 1)
+        {
+            return clusterInfos.get(0).getTimeSkew(instances);
+        }
+
+        Map<String, List<RingInstance>> instancesByClusterId = instances.stream().collect(Collectors.groupingBy(instance -> {
+            String clusterId = instance.clusterId();
+            Preconditions.checkState(clusterId != null,
+                                     "RingInstance must define its clusterId for coordinated write");
+            return clusterId;
+        }));
+        long localNow = System.currentTimeMillis();
+        long maxDiff = 0;
+        TimeSkewResponse largestSkew = null;
+        for (Map.Entry<String, List<RingInstance>> entry : instancesByClusterId.entrySet())
+        {
+            String clusterId = entry.getKey();
+            List<RingInstance> instancesOfCluster = entry.getValue();
+            ClusterInfo clusterInfo = cluster(clusterId);
+            Preconditions.checkState(clusterInfo != null, "ClusterInfo not found with clusterId: " + clusterId);
+            TimeSkewResponse response = clusterInfo.getTimeSkew(instancesOfCluster);
+            long d = Math.abs(response.currentTime - localNow);
+            if (Math.abs(response.currentTime - localNow) > maxDiff)
+            {
+                maxDiff = d;
+                largestSkew = response;
+            }
+        }
+        return largestSkew;
+    }
```

Comment:
If we add an `allInstances` method to `clusterInfo` this becomes _much simpler_:
```suggestion
    public TimeSkewResponse getTimeSkew(List<RingInstance> instances)
    {
        return applyOnEach(clusterInfo ->
                           clusterInfo.getTimeSkew(clusterInfo.allInstances()))
               .values()
               .stream()
               .max(Comparator.comparingLong(tsr -> Math.abs(tsr.currentTime - System.currentTimeMillis())))
               .orElseThrow(() -> new RuntimeException("Could not load time skew data from any cluster."));
    }
```

I'd say the overhead of the `applyOnEach` is worth the code clarity by removing the size check at the beginning, but up to you.

`allInstances` is a one-liner in `ClusterInfo`:

```java
    public List<RingInstance> allInstances()
    {
        return new ArrayList<>(getTokenRangeReplicas().allInstances());
    }
```

and in `ClusterInfoGroup`:

``` java
    public List<RingInstance> allInstances()
    {
        return applyOnEach(ci -> ci.allInstances())
               .values().stream().flatMap(Collection::stream)
               .collect(Collectors.toUnmodifiableList());
    }
```

Reply:
The suggested implementation is not identical.

Reply:
I think it makes better sense to drop the parameter in the method. Let me know.

```
public TimeSkewResponse timeSkew()
```

Reply:
I pushed https://github.com/apache/cassandra-analytics/pull/80/commits/45c8e0bebdfccd1a6071213556fbd986771a7a92 to address this comment